src/states/node.rs

```diff
@@ -653,56 +661,77 @@ impl Node {
         result.and(self.handle_signed_message(content, route, hop_name, &sent_to))
     }
 
-    // Acknowledge reception of the message and broadcast to our group if necessary
-    // The function is only called when we are in the destination authority
-    fn ack_and_broadcast(&mut self,
-                         signed_msg: &SignedMessage,
-                         route: u8,
-                         hop_name: XorName,
-                         sent_to: &BTreeSet<XorName>)
-                         -> Evented<()> {
-        let mut result = Evented::empty();
-        self.send_ack(signed_msg.routing_message(), route).extract(&mut result);
-        // If the destination is our group we need to forward it to the rest of the group
-        if signed_msg.routing_message().dst.is_multiple() {
-            if let Err(error) = self.send_signed_message(signed_msg, route, &hop_name, sent_to)
-                .extract(&mut result) {
-                debug!("{:?} Failed to send {:?}: {:?}", self, signed_msg, error);
-            }
-        }
-        result
-    }
-
+    // Verify the message, then, if it is for us, handle the enclosed routing message; if not,
+    // forward it.
+    //
+    // If we are the originator or accumulator of the message, then `hop_name` is our name;
+    // otherwise, it is the name of the node which forwarded us the message.
     fn handle_signed_message(&mut self,
-                             signed_msg: SignedMessage,
+                             mut signed_msg: SignedMessage,
                              route: u8,
                              hop_name: XorName,
                              sent_to: &BTreeSet<XorName>)
                              -> Evented<Result<(), RoutingError>> {
         let mut result = Evented::empty();
-        try_ev!(signed_msg.check_integrity(self.min_group_size()), result);
+
+        let hop_prefix = try_ev!(self.peer_mgr
+            .routing_table()
+            .find_group_prefix(&hop_name)
+            .ok_or(RoutingTableError::NoSuchPeer), result);
+        let section_list = if signed_msg.routing_message().src.is_client() ||
+                              self.in_authority(&signed_msg.routing_message().dst) {
+            // TODO: we should still verify the sending nodes when the sender is our own section
+            // or group, but doing so causes a lot of verification failures at the moment.
+            // (In any case don't call add_relaying_section if from our own section.)
+            None
+        } else {
+            let list = self.section_list_sigs.get_signed_list(&hop_prefix);
```

Comment:
Agree, except that (sometimes) we need the signed section list below anyway. Not sure which approach is best to be honest.